File: test/Claude/serial2WebSocket.cpp
Il n'y a pas encore de rerouting de la ligne série dans le code actuel — le terminal affiche uniquement les messages générés par le JavaScript lui-même (actions sur les sliders, joystick, etc.). Voici comment implémenter le vrai reroutage Serial → WebSocket → terminal web.

Le principe
ESP32                          Navigateur
──────                         ──────────
Serial.printf(...)  ──WS──►   ws.onmessage → addLog(...)
Le Serial normal écrit sur l'UART. Il faut intercepter ces writes et les dupliquer vers le WebSocket.








// ─────────────────────────────────────────────────────────
// Classe qui hérite de Print (base de Serial)
// et redirige chaque write() vers le WebSocket
// ─────────────────────────────────────────────────────────
class SerialOverWS : public Print {
public:
  AsyncWebSocket* _ws;
  String          _buf;   // buffer ligne courante

  SerialOverWS(AsyncWebSocket* ws) : _ws(ws) {}

  // Appelé pour CHAQUE caractère ou bloc écrit
  size_t write(uint8_t c) override {
    // On accumule dans le buffer jusqu'au '\n'
    if (c == '\n') {
      flush();
    } else {
      _buf += (char)c;
    }
    return 1;
  }

  size_t write(const uint8_t* buf, size_t size) override {
    for (size_t i = 0; i < size; i++) write(buf[i]);
    return size;
  }

  // Envoie la ligne accumulée à tous les clients WS
  void flush() {
    if (_buf.length() == 0) return;

    // Format JSON attendu par le terminal JS
    StaticJsonDocument<256> doc;
    doc["event"] = "serial";
    doc["msg"]   = _buf;

    String out;
    serializeJson(doc, out);
    _ws->textAll(out);   // broadcast à tous les clients connectés

    _buf = "";            // vide le buffer
  }
};

// Instance globale
SerialOverWS* serialWS = nullptr;

// Macro pratique : écrit sur Serial ET sur le WS
#define LOG(fmt, ...) {                            \
  Serial.printf(fmt, ##__VA_ARGS__);               \
  if (serialWS) serialWS->printf(fmt, ##__VA_ARGS__); \
}





Dans setup(), après avoir créé le WebSocket :
cppserialWS = new SerialOverWS(&ws);

// Maintenant on peut utiliser LOG() partout
LOG("[WIFI] Connecté → IP: %s\n", WiFi.localIP().toString().c_str());
LOG("[PID] TA=%.1f KP=%.3f KI=%.3f KD=%.3f\n", TA, KP, KI, KD);
LOG("[IMU] angle=%.2f° err=%.2f°\n", angle, error);

File: test/Claude/serial2WebSocket.js
ws.onmessage = e => {
  try {
    const msg = JSON.parse(e.data);

    if (msg.event === 'serial') {
      // Message venant de Serial/LOG() sur l'ESP32
      addLog('[ESP32] ' + msg.msg, '');
    }
    else if (msg.event === 'saved')  { ... }
    else if (msg.event === 'error')  { ... }
    // etc.
  } catch(_) {
    addLog('[WS] ' + e.data, '');
  }
};
```

---

## Flux complet, pas à pas
```
1. Code C++ appelle :
   LOG("[IMU] angle=89.7° err=0.3°\n")

2. SerialOverWS::write() reçoit chaque caractère
   → accumule dans _buf : "[IMU] angle=89.7° err=0.3°"
   → '\n' détecté → flush()

3. flush() sérialise en JSON :
   {"event":"serial","msg":"[IMU] angle=89.7° err=0.3°"}
   → ws.textAll() → envoyé à tous les navigateurs connectés

4. js ws.onmessage reçoit le JSON
   → msg.event === 'serial'
   → addLog('[ESP32] [IMU] angle=89.7° err=0.3°', '')

5. addLog() crée une <div class="term-line"> horodatée
   → insérée dans #terminal → scrollTop mis à jour
   
   
   
   Pourquoi passer par une classe Print ?
Serial sur Arduino/ESP32 est lui-même une classe qui hérite de Print. La classe Print définit une interface avec une seule méthode abstraite : write(uint8_t). Tout le reste (printf, println, print) est construit par-dessus. En héritant de Print et en redéfinissant uniquement write(), on intercepte toutes les formes d'écriture sans modifier le reste du code.